Objects/frameobject.c

```diff
@@ -2105,17 +2105,20 @@ PyFrame_GetGenerator(PyFrameObject *frame)
 PyObject*
 _PyEval_BuiltinsFromGlobals(PyThreadState *tstate, PyObject *globals)
 {
-    PyObject *builtins = PyDict_GetItemWithError(globals, &_Py_ID(__builtins__));
-    if (builtins) {
+    PyObject *builtins;
+    int has_builtins = PyMapping_GetOptionalItem(
+        globals, &_Py_ID(__builtins__), &builtins);
+    if (has_builtins < 0) {
+        return NULL;
+    }
+    if (has_builtins) {
+        /* release reference right away since we mean to only borrow it */
```

Comment:
This is not safe. If you want to go this route, you should make `_PyEval_BuiltinsFromGlobals()` returning a non-borrowed reference.

But I think that this feature requires too many changes. It is dangerous to do this without very good reasons.

Reply:
Well spotted. Thank you so much for the tips!

Reply:
I personally think the OP of the reported issue has a good use case for `globals()` to a mapping, for the same reason why we allow `locals()` to be a mapping. It just makes the namespace more customizable. As long as those who customize the namespace are aware of what they are doing I don't see any harm in loosening the restrictions since the default namespace continues to be a dict.